services/gallery.js

```diff
@@ -0,0 +1,55 @@
+class GalleryService {
+	constructor(api, httpParamSerializer, tagFilter, userId) {
+		this.api = api;
+		this.serializer = httpParamSerializer;
+		this.videos = [];
+		this.queryParams = {
+			order: '-date',
+			offset: 0,
+			limit: 10
+		};
+		if (tagFilter instanceof Array && tagFilter.length > 0) {
+			this.createTagFilter(tagFilter);
+		}
+		if (userId == parseInt(userId, 10)) {
+			this.userId = parseInt(userId, 10);
+		}
+	}
+
+	createTagFilter(tagFilter) {
+		this.queryParams.tag = tagFilter.filter(item => !item.startsWith('-'));
+		this.queryParams.excludeTag = tagFilter.filter(item => item.startsWith('-')).map(item => item.substring(1));
+	}
+
+	getQuery() {
+		let baseQuery = "/video?";
+		if (this.userId !== undefined) {
+			baseQuery = '/user/' + this.userId + baseQuery;
+		}
+		return baseQuery + this.serializer(this.queryParams);
+	}
+
+	getVideoList(cb) {
+		console.log("query url is ", this.api.url + this.getQuery());
+		this.api.get(this.api.url + this.getQuery(), function (data, status) {
+			let success = false;
+			if (status < 400 && data.length > 0) {
+				this.videos = this.videos.concat(data);
+				this.queryParams.offset += data.length;
+				success = true;
+			}
+			cb(success);
+		}.bind(this));
+	}
+}
+
+function galleryServiceGenerator(api, $httpParamSerializer) {
+	return {
+		getInstance: function (tagFilter, userId) {
+			return new GalleryService(api, $httpParamSerializer, tagFilter, userId);
```

Comment:
Services are singletons, but we're using a singleton to instanciate new modules, which is not the idea of services. To do that, angular have other patterns as the own directives, or even controllers. In this case, I think we coud find a better way to do this.

Reply:
I understand know why it would be hard to find examples of factories in angular. So factory in angular is not intended to build factories?

Current "service" implementation makes sense to me with current services developed, such as api or user services, that makes sense to be singletons. What I tried to make here is another thing on top of a service, the gallery factory, that, using api service, will provide video lists paginated support, and can then generalize existing code for easier reusing (such as in exisiting gallery: one instance for handling featured videos, another for handling non featured separately) and better maintainability.

So I will need help choosing the best angular construct to implement this approach